contracts/core/MarginPoolV2.sol

```diff
@@ -0,0 +1,244 @@
+     */
+    function setBuyerWhitelistedStatus(address _buyer, bool _whitelisted) external onlyOwner {
+        require(_buyer != address(0), "!_buyer");
+
+        whitelistedOTokenBuyer[_buyer] = _whitelisted;
+        emit SetOTokenBuyerWhitelist(_buyer, _whitelisted);
+    }
+
+    /**
+     * @notice Set Borrow percent of collateral asset
+     * @param _collateral address of collateral asset
+     * @param _borrowPCT borrow PCT
+     */
+    function setBorrowPCT(address _collateral, uint256 _borrowPCT) external onlyOwner {
+        borrowPCT[_collateral] = _borrowPCT;
+        emit SetBorrowPCT(_collateral, _borrowPCT);
+    }
+
+    /**
+     * @notice Lends out asset to market maker
+     * @param _oToken address of the oToken laying claim to the collateral assets
+     * @param _amount amount of the asset to lend to borrower
+     */
+    function borrow(address _oToken, uint256 _amount) public onlyWhitelistedBorrower {
+        require(
+            WhitelistInterface(AddressBookInterface(addressBook).getWhitelist()).isWhitelistedOtoken(_oToken),
+            "MarginPool: oToken is not whitelisted"
+        );
+
+        OtokenInterface oToken = OtokenInterface(_oToken);
+
+        address collateralAsset = oToken.collateralAsset();
+        uint256 outstandingAssetBorrow = borrowed[msg.sender][collateralAsset];
+
+        require(!oToken.isPut(), "MarginPool: oToken is not a call option");
+
+        require(borrowPCT[collateralAsset] > 0, "MarginPool: Borrowing is paused for collateral asset");
+
+        // Make sure borrow does not attempt to borrow with an expired oToken
+        require(
+            oToken.expiryTimestamp() > block.timestamp,
+            "MarginPool: Cannot borrow collateral asset of expired oToken"
+        );
+
+        uint256 collateralAssetDecimals = ERC20Interface(collateralAsset).decimals();
+
+        // Each oToken represents 1:1 of collateral token
+        // So a market maker can borrow at most our oToken balance in the collateral asset
+        uint256 collateralAllocatedToMM = ERC20Interface(_oToken).balanceOf(msg.sender);
+        if (collateralAssetDecimals >= 8) {
+            collateralAllocatedToMM = collateralAllocatedToMM.mul(10**(collateralAssetDecimals.sub(8)));
+        } else {
+            collateralAllocatedToMM = collateralAllocatedToMM.div(10**(uint256(8).sub(collateralAssetDecimals)));
+        }
+
+        require(
+            _amount <=
+                collateralAllocatedToMM.mul(borrowPCT[collateralAsset]).div(TOTAL_PCT).sub(outstandingAssetBorrow),
+            "MarginPool: Borrowing more than allocated"
+        );
+
+        borrowed[msg.sender][collateralAsset] = outstandingAssetBorrow.add(_amount);
+
+        // transfer _asset _amount from pool to borrower
+        ERC20Interface(collateralAsset).safeTransfer(msg.sender, _amount);
+        emit Borrow(_oToken, collateralAsset, _amount, msg.sender);
```

Comment:
Two tricks that would probably mess up the accounting: 
1. whitelisted parties can collude and make others unable to borrow: they can transfer the oToken balance back and forth among whitelisted parties and borrow more than what they should have.
2. whitelisted parties can flash mint bunch of oTokens and borrow almost every borrowable asset from the pool.

I think the pool should lock up oToken when someone come and borrow, to prevent these scenarios

Reply:
Agree

Reply:
Scenario #1 is possible. I just chalked it up to "master loan agreement will be signed so there is no incentive here" but locking up oTokens definitely solves this. 

Scenario #2 flashminting is solved because whereas before they could flash mint whitelisted oToken and borrow underlying from pool and then just burn the original oTokens - with a lock they are basically swapping their own collateral with that of the pools so there is no gain here.

Reply:
@antoncoding @haythem96 thanks for the thorough review